[BUG] java.lang.ClassCastException: SerializeConcatHostBuffersDeserializeBatch cannot be cast to.HashedRelation

**Describe the bug**
User application hit the following exception on Databricks 10.4:

```

Py4JJavaError: An error occurred while calling o530.javaToPython.
: org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:446)
	at org.apache.spark.sql.execution.SubqueryExec.executeCollect(basicPhysicalOperators.scala:953)
	at org.apache.spark.sql.execution.InSubqueryExec.updateResult(subquery.scala:137)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$waitForSubqueries$1(SparkPlan.scala:308)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$waitForSubqueries$1$adapted(SparkPlan.scala:307)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.sql.execution.SparkPlan.waitForSubqueries(SparkPlan.scala:307)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:269)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at com.nvidia.spark.rapids.GpuFilterExec.doExecuteColumnar(basicPhysicalOperators.scala:466)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at com.nvidia.spark.rapids.GpuCoalesceBatches.doExecuteColumnar(GpuCoalesceBatches.scala:713)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD$lzycompute(GpuShuffleExchangeExecBase.scala:221)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD(GpuShuffleExchangeExecBase.scala:221)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBaseWithMetrics.mapOutputStatisticsFuture$lzycompute(GpuShuffleExchangeExecBase.scala:141)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBaseWithMetrics.mapOutputStatisticsFuture(GpuShuffleExchangeExecBase.scala:140)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.$anonfun$submitShuffleJob$1(ShuffleExchangeExec.scala:73)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.submitShuffleJob(ShuffleExchangeExec.scala:73)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.submitShuffleJob$(ShuffleExchangeExec.scala:72)
	at org.apache.spark.rapids.shims.GpuShuffleExchangeExec.submitShuffleJob(GpuShuffleExchangeExec.scala:27)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.shuffleFuture$lzycompute(QueryStageExec.scala:272)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.shuffleFuture(QueryStageExec.scala:272)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.doMaterialize(QueryStageExec.scala:274)
	at org.apache.spark.sql.execution.adaptive.QueryStageExec.materialize(QueryStageExec.scala:123)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$5(AdaptiveSparkPlanExec.scala:400)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$5$adapted(AdaptiveSparkPlanExec.scala:398)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$1(AdaptiveSparkPlanExec.scala:398)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:968)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.getFinalPhysicalPlan(AdaptiveSparkPlanExec.scala:370)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.withFinalPlanUpdate(AdaptiveSparkPlanExec.scala:502)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.doExecute(AdaptiveSparkPlanExec.scala:513)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:226)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:222)
	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:253)
	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:252)
	at org.apache.spark.sql.Dataset.javaToPython(Dataset.scala:3731)
	at sun.reflect.GeneratedMethodAccessor858.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:295)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:251)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.util.concurrent.ExecutionException: org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:438)
	... 70 more
Caused by: org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:446)
	at org.apache.spark.sql.execution.SubqueryBroadcastExec.executeCollect(SubqueryBroadcastExec.scala:144)
	at org.apache.spark.sql.execution.InSubqueryExec.updateResult(subquery.scala:137)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$waitForSubqueries$1(SparkPlan.scala:308)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$waitForSubqueries$1$adapted(SparkPlan.scala:307)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.sql.execution.SparkPlan.waitForSubqueries(SparkPlan.scala:307)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:269)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at com.nvidia.spark.rapids.GpuFilterExec.doExecuteColumnar(basicPhysicalOperators.scala:466)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at com.nvidia.spark.rapids.GpuCoalesceBatches.doExecuteColumnar(GpuCoalesceBatches.scala:713)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinExecBase.doColumnarBroadcastJoin(GpuBroadcastHashJoinExecBase.scala:175)
	at org.apache.spark.sql.rapids.execution.GpuBroadcastHashJoinExecBase.doExecuteColumnar(GpuBroadcastHashJoinExecBase.scala:192)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at com.nvidia.spark.rapids.GpuProjectExec.doExecuteColumnar(basicPhysicalOperators.scala:172)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:253)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:249)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD$lzycompute(GpuShuffleExchangeExecBase.scala:221)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD(GpuShuffleExchangeExecBase.scala:221)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBaseWithMetrics.mapOutputStatisticsFuture$lzycompute(GpuShuffleExchangeExecBase.scala:141)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBaseWithMetrics.mapOutputStatisticsFuture(GpuShuffleExchangeExecBase.scala:140)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.$anonfun$submitShuffleJob$1(ShuffleExchangeExec.scala:73)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:270)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:266)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.submitShuffleJob(ShuffleExchangeExec.scala:73)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeLike.submitShuffleJob$(ShuffleExchangeExec.scala:72)
	at org.apache.spark.rapids.shims.GpuShuffleExchangeExec.submitShuffleJob(GpuShuffleExchangeExec.scala:27)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.shuffleFuture$lzycompute(QueryStageExec.scala:272)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.shuffleFuture(QueryStageExec.scala:272)
	at org.apache.spark.sql.execution.adaptive.ShuffleQueryStageExec.doMaterialize(QueryStageExec.scala:274)
	at org.apache.spark.sql.execution.adaptive.QueryStageExec.materialize(QueryStageExec.scala:123)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$5(AdaptiveSparkPlanExec.scala:400)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$5$adapted(AdaptiveSparkPlanExec.scala:398)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.$anonfun$getFinalPhysicalPlan$1(AdaptiveSparkPlanExec.scala:398)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:968)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.getFinalPhysicalPlan(AdaptiveSparkPlanExec.scala:370)
	at org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanExec.withFinalPlanUpdate(AdaptiveSparkPlanExec.scala:502)
	at org.apache.spark.sql.execution.ResultCacheManager.computeResult(ResultCacheManager.scala:580)
	at org.apache.spark.sql.execution.ResultCacheManager.$anonfun$getOrComputeResultInternal$1(ResultCacheManager.scala:528)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResultInternal(ResultCacheManager.scala:527)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:424)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:403)
	at org.apache.spark.sql.execution.SparkPlan.executeCollectResult(SparkPlan.scala:424)
	at org.apache.spark.sql.execution.SparkPlan.executeCollect(SparkPlan.scala:400)
	at org.apache.spark.sql.execution.SubqueryExec.$anonfun$relationFuture$1(basicPhysicalOperators.scala:929)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:968)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$4(SQLExecution.scala:448)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$3(SQLExecution.scala:448)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$2(SQLExecution.scala:447)
	at org.apache.spark.sql.execution.SQLExecution$.withOptimisticTransaction(SQLExecution.scala:465)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withThreadLocalCaptured$1(SQLExecution.scala:446)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.$anonfun$run$1(SparkThreadLocalForwardingThreadPoolExecutor.scala:104)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:68)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingHelper.runWithCaptured$(SparkThreadLocalForwardingThreadPoolExecutor.scala:54)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.runWithCaptured(SparkThreadLocalForwardingThreadPoolExecutor.scala:101)
	at org.apache.spark.util.threads.SparkThreadLocalCapturingRunnable.run(SparkThreadLocalForwardingThreadPoolExecutor.scala:104)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more
Caused by: java.util.concurrent.ExecutionException: java.lang.ClassCastException: org.apache.spark.sql.rapids.execution.SerializeConcatHostBuffersDeserializeBatch cannot be cast to org.apache.spark.sql.execution.joins.HashedRelation
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:438)
	... 91 more
Caused by: java.lang.ClassCastException: org.apache.spark.sql.rapids.execution.SerializeConcatHostBuffersDeserializeBatch cannot be cast to org.apache.spark.sql.execution.joins.HashedRelation
	at org.apache.spark.sql.execution.SubqueryBroadcastExec.$anonfun$relationFuture$1(SubqueryBroadcastExec.scala:111)
	... 18 more
```

--- a/integration_tests/src/main/python/delta_zorder_test.py
+++ b/integration_tests/src/main/python/delta_zorder_test.py
@@ -1,4 +1,4 @@
-# Copyright (c) 2022, NVIDIA CORPORATION.
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/integration_tests/src/main/python/delta_zorder_test.py
+++ b/integration_tests/src/main/python/delta_zorder_test.py
@@ -125,7 +125,7 @@ def test_delta_dfp_reuse_broadcast_exchange(spark_tmp_table_factory, s_index, aq
             .mode("overwrite") \
             .partitionBy("key", "skey") \
             .saveAsTable(fact_table)
-        spark.sql("OPTIMIZE {} ZORDER BY ex_key".format(fact_table))
+        spark.sql("OPTIMIZE {} ZORDER BY ex_key".format(fact_table)).show()
 
         df = gen_df(spark, [
             ('key', LongGen(nullable=False, min_val=0, max_val=9, special_cases=[])),

--- a/integration_tests/src/main/python/delta_zorder_test.py
+++ b/integration_tests/src/main/python/delta_zorder_test.py
@@ -156,6 +156,7 @@ def test_delta_dfp_reuse_broadcast_exchange(spark_tmp_table_factory, s_index, aq
         exist_classes,
         # Ensure Dynamic File Pruning kicks in by setting thresholds to 0
         conf=dict(_exchange_reuse_conf + [
+            ('spark.databricks.optimizer.dynamicFilePruning', 'true'),
             ('spark.databricks.optimizer.deltaTableSizeThreshold', '0'),
             ('spark.databricks.optimizer.deltaTableFilesThreshold', '0'),
             ('spark.sql.adaptive.enabled', aqe_enabled)]))

--- a/integration_tests/src/main/python/delta_zorder_test.py
+++ b/integration_tests/src/main/python/delta_zorder_test.py
@@ -14,10 +14,12 @@
 
 import pytest
 
-from asserts import assert_gpu_and_cpu_are_equal_collect
+from asserts import assert_cpu_and_gpu_are_equal_collect_with_capture, assert_gpu_and_cpu_are_equal_collect
 from data_gen import *
 from marks import allow_non_gpu, ignore_order, delta_lake
-from spark_session import is_databricks_runtime, with_cpu_session, with_gpu_session
+from spark_session import is_databricks_runtime, with_cpu_session, with_gpu_session, is_databricks104_or_later, is_databricks113_or_later
+
+from dpp_test import _exchange_reuse_conf
 
 # Almost all of this is the metadata query
 # the important part is to not have InterleaveBits or HilbertLongIndex and PartitionerExpr
@@ -48,3 +50,112 @@ def test_delta_zorder(spark_tmp_table_factory):
                   "spark.rapids.sql.castFloatToString.enabled": True,
                   "spark.rapids.sql.explain": "ALL"})
 
+_statements = [
+    # join on z-ordered column
+    '''
+    SELECT fact.ex_key, sum(fact.value)
+    FROM {0} fact
+    JOIN {1} dim
+    ON fact.ex_key = dim.ex_key
+    WHERE dim.filter = {2}
+    GROUP BY fact.ex_key
+    ''',
+    # join on 1 partitioned and 1 z-ordered column
+    '''
+    SELECT fact.key, fact.ex_key, sum(fact.value)
+    FROM {0} fact
+    JOIN {1} dim
+    ON fact.key = dim.key AND fact.ex_key = dim.ex_key
+    WHERE dim.filter = {2}
+    GROUP BY fact.key, fact.ex_key
+    ''',
+    # join on 2 partitioned and 1 z-ordered columns
+    '''
+    SELECT fact.key, fact.skey, fact.ex_key, sum(fact.value)
+    FROM {0} fact
+    JOIN {1} dim
+    ON fact.key = dim.key AND fact.skey = dim.skey AND fact.ex_key = dim.ex_key
+    WHERE dim.filter = {2}
+    GROUP BY fact.key, fact.skey, fact.ex_key
+    ''',
+    # reused subquery, join on z-ordered column
+    '''
+    SELECT ex_key, max(value)
+    FROM (
+        SELECT fact.ex_key as ex_key, fact.value as value
+        FROM {0} fact
+        JOIN {1} dim
+        ON fact.ex_key = dim.ex_key
+        WHERE dim.filter = {2}
+    UNION ALL
+        SELECT fact.ex_key as ex_key, fact.value as value
+        FROM {0} fact
+        JOIN {1} dim
+        ON fact.ex_key = dim.ex_key
+        WHERE dim.filter = {2}
+    )
+    GROUP BY ex_key
+    '''
+]
+
+# This tests Dynamic File Pruning, a feature in Databricks that is similar to Dynamic Partition Pruning 
+# except that it adds the DynamicPruningExpression for columns that are not partition columns but are still
+# optimized. In this case the DynamicPruningExpression should be added to the DataFilters in the scan.
+# This test is very similar to `test_dpp_reuse_broadcast_exchange` but it tests joining using a Z-ordered
+# column
+@delta_lake
+@ignore_order(local=True)
+@pytest.mark.skipif(not is_databricks104_or_later(), reason="Dynamic File Pruning is only supported in Databricks 10.4+")
+@pytest.mark.parametrize('s_index', list(range(len(_statements))), ids=idfn)
+@pytest.mark.parametrize('aqe_enabled', ['false', 'true'])
+def test_delta_dfp_reuse_broadcast_exchange(spark_tmp_table_factory, s_index, aqe_enabled):
+    fact_table, dim_table = spark_tmp_table_factory.get(), spark_tmp_table_factory.get()
+
+    def build_and_optimize_tables(spark):
+        # Note that ex_key is a high-cardinality column, which makes it a good candidate for 
+        # for Z-ordering, which then means it can then be used in Dynamic File Pruning in joins
+        df = gen_df(spark, [
+            ('key', LongGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
+            ('skey', LongGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
+            ('ex_key', LongGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('value', int_gen),
+        ], 10000)
+
+        df.write.format("delta") \
+            .mode("overwrite") \
+            .partitionBy("key", "skey") \
+            .saveAsTable(fact_table)
+        spark.sql("OPTIMIZE {} ZORDER BY ex_key".format(fact_table))
+
+        df = gen_df(spark, [
+            ('key', LongGen(nullable=False, min_val=0, max_val=9, special_cases=[])),
+            ('skey', LongGen(nullable=False, min_val=0, max_val=4, special_cases=[])),
+            ('ex_key', LongGen(nullable=False, min_val=0, max_val=10000, special_cases=[])),
+            ('value', int_gen),
+            ('filter', RepeatSeqGen(
+                IntegerGen(min_val=0, max_val=2000, special_cases=[]), length=2000 // 20))
+        ], 2000)
+        df.write.format("delta") \
+            .mode("overwrite") \
+            .saveAsTable(dim_table)
+        return df.select('filter').first()[0]
+
+    filter_val = with_cpu_session(build_and_optimize_tables)
+
+    statement = _statements[s_index].format(fact_table, dim_table, filter_val)
+
+    if is_databricks113_or_later() and aqe_enabled == 'true':
+        # SubqueryBroadcastExec is unoptimized in Databricks 11.3 with EXECUTOR_BROADCAST
+        # See https://github.com/NVIDIA/spark-rapids/issues/7425
+        exist_classes='DynamicPruningExpression,SubqueryBroadcastExec,ReusedExchangeExec'
+    else:
+        exist_classes='DynamicPruningExpression,GpuSubqueryBroadcastExec,ReusedExchangeExec'
+    assert_cpu_and_gpu_are_equal_collect_with_capture(
+        lambda spark: spark.sql(statement),
+        # The existence of GpuSubqueryBroadcastExec indicates the reuse works on the GPU
+        exist_classes,
+        # Ensure Dynamic File Pruning kicks in by setting thresholds to 0
+        conf=dict(_exchange_reuse_conf + [
+            ('spark.databricks.optimizer.deltaTableSizeThreshold', '0'),
+            ('spark.databricks.optimizer.deltaTableFilesThreshold', '0'),
+            ('spark.sql.adaptive.enabled', aqe_enabled)]))

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
@@ -19,7 +19,7 @@ package com.nvidia.spark.rapids.shims
 import com.nvidia.spark.rapids._
 
 import org.apache.spark.internal.Logging
-import org.apache.spark.sql.catalyst.expressions.DynamicPruningExpression
+import org.apache.spark.sql.catalyst.expressions.{DynamicPruningExpression, Expression}
 import org.apache.spark.sql.execution._
 import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InMemoryFileIndex}
 import org.apache.spark.sql.execution.datasources.json.JsonFileFormat
@@ -81,41 +81,27 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
     }
   }
 
+  private def convertDynamicPruningFilters(filters: Seq[Expression]): Seq[Expression] = {
+    filters.map { filter =>
+      filter.transformDown {
+        case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
+          inSub.plan match {
+            case bc: SubqueryBroadcastExec =>
+              dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
+            case reuse @ ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
+              dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
+            case _ =>
+              dpe
+          }
+      }
+    }
+  }
 
   // Support partitionFilters in Dynamic Partition Pruning
-  private lazy val partitionFilters = {
-    wrapped.partitionFilters.map { filter =>
-      filter.transformDown {
-        case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
-          inSub.plan match {
-            case bc: SubqueryBroadcastExec =>
-              dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
-            case reuse @ ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
-              dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
-            case _ =>
-              dpe
-          }
-      }
-    }
-  }
+  private lazy val partitionFilters = convertDynamicPruningFilters(wrapped.partitionFilters)
 
   // Support dataFilters in Dynamic File Pruning
-  private lazy val dataFilters = {
-    wrapped.dataFilters.map { filter =>
-      filter.transformDown {
-        case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
-          inSub.plan match {
-            case bc: SubqueryBroadcastExec =>
-              dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
-            case reuse @ ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
-              dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
-            case _ =>
-              dpe
-          }
-      }
-    }
-  }
-
+  private lazy val dataFilters = convertDynamicPruningFilters(wrapped.dataFilters)
 
   // partition filters and data filters are not run on the GPU
   override val childExprs: Seq[ExprMeta[_]] = Seq.empty

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
@@ -18,6 +18,7 @@ package com.nvidia.spark.rapids.shims
 
 import com.nvidia.spark.rapids._
 
+import org.apache.spark.internal.Logging
 import org.apache.spark.sql.catalyst.expressions.DynamicPruningExpression
 import org.apache.spark.sql.execution._
 import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, InMemoryFileIndex}
@@ -30,7 +31,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
     conf: RapidsConf,
     parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule)
-    extends SparkPlanMeta[FileSourceScanExec](plan, conf, parent, rule) {
+    extends SparkPlanMeta[FileSourceScanExec](plan, conf, parent, rule) with Logging {
 
   // Replaces SubqueryBroadcastExec inside dynamic pruning filters with GPU counterpart
   // if possible. Instead regarding filters as childExprs of current Meta, we create
@@ -42,6 +43,12 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
     val convertBroadcast = (bc: SubqueryBroadcastExec) => {
       val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
       meta.tagForExplain()
+      if (conf.shouldExplain) {
+        val explain = meta.explain(conf.shouldExplainAll)
+        if (explain.nonEmpty) {
+          logWarning(s"\n$explain")
+        }
+      }
       val converted = meta.convertIfNeeded()
       // Because the PlanSubqueries rule is not called (and does not work as expected),
       // we might actually have to fully convert the subquery plan as the plugin would

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
@@ -39,48 +39,49 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
   // FileSourceScan is independent from the replacement of the partitionFilters. It is
   // possible that the FileSourceScan is on the CPU, while the dynamic partitionFilters
   // are on the GPU. And vice versa.
-  private lazy val partitionFilters = {
-    val convertBroadcast = (bc: SubqueryBroadcastExec) => {
-      val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
-      meta.tagForExplain()
-      if (conf.shouldExplain) {
-        val explain = meta.explain(conf.shouldExplainAll)
-        if (explain.nonEmpty) {
-          logWarning(s"\n$explain")
-        }
-      }
-      val converted = meta.convertIfNeeded()
-      // Because the PlanSubqueries rule is not called (and does not work as expected),
-      // we might actually have to fully convert the subquery plan as the plugin would
-      // intend (in this case calling GpuTransitionOverrides to insert GpuCoalesceBatches,
-      // etc.) to match the other side of the join to reuse the BroadcastExchange.
-      // This happens when SubqueryBroadcast has the original (Gpu)BroadcastExchangeExec
-      converted match {
-        case e: GpuSubqueryBroadcastExec => e.child match {
-          // If the GpuBroadcastExchange is here, then we will need to run the transition
-          // overrides here
-          case _: GpuBroadcastExchangeExec =>
-            var updated = ApplyColumnarRulesAndInsertTransitions(Seq(), true)
-                .apply(converted)
-            updated = (new GpuTransitionOverrides()).apply(updated)
-            updated match {
-              case h: GpuBringBackToHost =>
-                h.child.asInstanceOf[BaseSubqueryExec]
-              case c2r: GpuColumnarToRowExec =>
-                c2r.child.asInstanceOf[BaseSubqueryExec]
-              case _: GpuSubqueryBroadcastExec =>
-                updated.asInstanceOf[BaseSubqueryExec]
-            }
-          // Otherwise, if this SubqueryBroadcast is using a ReusedExchange, then we don't
-          // do anything further
-          case _: ReusedExchangeExec =>
-            converted.asInstanceOf[BaseSubqueryExec]
-        }
-        case _ =>
-          converted.asInstanceOf[BaseSubqueryExec]
+  private def convertBroadcast(bc: SubqueryBroadcastExec):BaseSubqueryExec = {
+    val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
+    meta.tagForExplain()
+    if (conf.shouldExplain) {
+      val explain = meta.explain(conf.shouldExplainAll)
+      if (explain.nonEmpty) {
+        logWarning(s"\n$explain")
       }
     }
+    val converted = meta.convertIfNeeded()
+    // Because the PlanSubqueries rule is not called (and does not work as expected),
+    // we might actually have to fully convert the subquery plan as the plugin would
+    // intend (in this case calling GpuTransitionOverrides to insert GpuCoalesceBatches,
+    // etc.) to match the other side of the join to reuse the BroadcastExchange.
+    // This happens when SubqueryBroadcast has the original (Gpu)BroadcastExchangeExec
+    converted match {
+      case e: GpuSubqueryBroadcastExec => e.child match {
+        // If the GpuBroadcastExchange is here, then we will need to run the transition
+        // overrides here
+        case _: GpuBroadcastExchangeExec =>
+          var updated = ApplyColumnarRulesAndInsertTransitions(Seq(), true)
+              .apply(converted)
+          updated = (new GpuTransitionOverrides()).apply(updated)
+          updated match {
+            case h: GpuBringBackToHost =>
+              h.child.asInstanceOf[BaseSubqueryExec]
+            case c2r: GpuColumnarToRowExec =>
+              c2r.child.asInstanceOf[BaseSubqueryExec]
+            case _: GpuSubqueryBroadcastExec =>
+              updated.asInstanceOf[BaseSubqueryExec]
+          }
+        // Otherwise, if this SubqueryBroadcast is using a ReusedExchange, then we don't
+        // do anything further
+        case _: ReusedExchangeExec =>
+          converted.asInstanceOf[BaseSubqueryExec]
+      }
+      case _ =>
+        converted.asInstanceOf[BaseSubqueryExec]
+    }
+  }
 
+
+  private lazy val partitionFilters = {
     wrapped.partitionFilters.map { filter =>
       filter.transformDown {
         case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
@@ -95,6 +96,23 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
       }
     }
   }
+
+  private lazy val dataFilters = {
+    wrapped.dataFilters.map { filter =>
+      filter.transformDown {
+        case dpe @ DynamicPruningExpression(inSub: InSubqueryExec) =>
+          inSub.plan match {
+            case bc: SubqueryBroadcastExec =>
+              dpe.copy(inSub.copy(plan = convertBroadcast(bc)))
+            case reuse @ ReusedSubqueryExec(bc: SubqueryBroadcastExec) =>
+              dpe.copy(inSub.copy(plan = reuse.copy(convertBroadcast(bc))))
+            case _ =>
+              dpe
+          }
+      }
+    }
+  }
+
 
   // partition filters and data filters are not run on the GPU
   override val childExprs: Seq[ExprMeta[_]] = Seq.empty
@@ -113,7 +131,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
   }
 
   override def convertToCpu(): SparkPlan = {
-    wrapped.copy(partitionFilters = partitionFilters)
+    wrapped.copy(partitionFilters = partitionFilters, dataFilters = dataFilters)
   }
 
   override def convertToGpu(): GpuExec = {
@@ -147,7 +165,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
             conf,
             wrapped.relation,
             partitionFilters,
-            wrapped.dataFilters)
+            dataFilters)
         } else {
           // convert time algorithm and read large files
           (wrapped.relation.location, None)
@@ -173,7 +191,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
       wrapped.optionalBucketSet,
       // TODO: Does Databricks have coalesced bucketing implemented?
       None,
-      wrapped.dataFilters,
+      dataFilters,
       wrapped.tableIdentifier,
       wrapped.disableBucketedScan,
       queryUsesInputFile = false,

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
@@ -38,7 +38,8 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
   // a new meta for SubqueryBroadcastExec. The reason is that the GPU replacement of
   // FileSourceScan is independent from the replacement of the partitionFilters. It is
   // possible that the FileSourceScan is on the CPU, while the dynamic partitionFilters
-  // are on the GPU. And vice versa.
+  // are on the GPU. And vice versa. The same applies for dataFilters in the case of
+  // Dynamic File Pruning
   private def convertBroadcast(bc: SubqueryBroadcastExec):BaseSubqueryExec = {
     val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
     meta.tagForExplain()
@@ -81,6 +82,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
   }
 
 
+  // Support partitionFilters in Dynamic Partition Pruning
   private lazy val partitionFilters = {
     wrapped.partitionFilters.map { filter =>
       filter.transformDown {
@@ -97,6 +99,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
     }
   }
 
+  // Support dataFilters in Dynamic File Pruning
   private lazy val dataFilters = {
     wrapped.dataFilters.map { filter =>
       filter.transformDown {

--- a/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
+++ b/sql-plugin/src/main/321+-db/scala/com/nvidia/spark/rapids/shims/FileSourceScanExecMeta.scala
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, NVIDIA CORPORATION.
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -40,7 +40,7 @@ class FileSourceScanExecMeta(plan: FileSourceScanExec,
   // possible that the FileSourceScan is on the CPU, while the dynamic partitionFilters
   // are on the GPU. And vice versa. The same applies for dataFilters in the case of
   // Dynamic File Pruning
-  private def convertBroadcast(bc: SubqueryBroadcastExec):BaseSubqueryExec = {
+  private def convertBroadcast(bc: SubqueryBroadcastExec): BaseSubqueryExec = {
     val meta = GpuOverrides.wrapAndTagPlan(bc, conf)
     meta.tagForExplain()
     if (conf.shouldExplain) {

